File: easycontrol/server/src/main/java/top/saymzx/easycontrol/server/helper/AudioEncode.java
/*
 * 本项目大量借鉴学习了开源投屏软件：Scrcpy，在此对该项目表示感谢
 */
package top.saymzx.easycontrol.server.helper;

import android.media.AudioRecord;
import android.media.MediaCodec;
import android.media.MediaCodecInfo;
import android.media.MediaFormat;
import android.os.Build;

import java.io.IOException;
import java.nio.ByteBuffer;

import top.saymzx.easycontrol.server.Server;
import top.saymzx.easycontrol.server.entity.Options;

public final class AudioEncode {
  private static MediaCodec encedec;
  private static AudioRecord audioCapture;
  private static boolean useOpus;

  public static boolean init() throws IOException {
    useOpus = Options.supportOpus && EncodecTools.isSupportOpus();
    byte[] bytes = new byte[]{0};
    try {
      // 从安卓12开始支持音频
      if (!Options.isAudio || Build.VERSION.SDK_INT < Build.VERSION_CODES.S) throw new Exception("版本低");
      setAudioEncodec();
      encedec.start();
      audioCapture = AudioCapture.init();
    } catch (Exception ignored) {
      Server.writeMain(ByteBuffer.wrap(bytes));
      return false;
    }
    bytes[0] = 1;
    Server.writeMain(ByteBuffer.wrap(bytes));
    bytes[0] = (byte) (useOpus ? 1 : 0);
    Server.writeMain(ByteBuffer.wrap(bytes));
    return true;
  }

  private static void setAudioEncodec() throws IOException {
    String codecMime = useOpus ? MediaFormat.MIMETYPE_AUDIO_OPUS : MediaFormat.MIMETYPE_AUDIO_AAC;
    encedec = MediaCodec.createEncoderByType(codecMime);
    MediaFormat encodecFormat = MediaFormat.createAudioFormat(codecMime, AudioCapture.SAMPLE_RATE, AudioCapture.CHANNELS);
    encodecFormat.setInteger(MediaFormat.KEY_BIT_RATE, 128000);
    encodecFormat.setInteger(MediaFormat.KEY_MAX_INPUT_SIZE, AudioCapture.AUDIO_PACKET_SIZE);
    encedec.configure(encodecFormat, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE);
  }

  public static void encodeIn() {
    try {
      int inIndex;
      do inIndex = encedec.dequeueInputBuffer(-1); while (inIndex < 0);
      ByteBuffer buffer = encedec.getInputBuffer(inIndex);
      if (buffer == null) return;
      int size = Math.min(buffer.remaining(), AudioCapture.AUDIO_PACKET_SIZE);
      audioCapture.read(buffer, size);
      encedec.queueInputBuffer(inIndex, 0, size, 0, 0);
    } catch (IllegalStateException ignored) {
    }
  }

  private static final MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();

  public static void encodeOut() throws IOException {
    try {
      // 找到已完成的输出缓冲区
      int outIndex;
      do outIndex = encedec.dequeueOutputBuffer(bufferInfo, -1); while (outIndex < 0);
      ByteBuffer buffer = encedec.getOutputBuffer(outIndex);
      if (buffer == null) return;
      if (useOpus) {
        if ((bufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) != 0) {
          buffer.getLong();
          int size = (int) buffer.getLong();
          buffer.limit(buffer.position() + size);
        }
        // 当无声音时不发送
        if (buffer.remaining() < 5) {
          encedec.releaseOutputBuffer(outIndex, false);
          return;
        }
      }
      ControlPacket.sendAudioEvent(buffer);
      encedec.releaseOutputBuffer(outIndex, false);
    } catch (IllegalStateException ignored) {
    }
  }

  public static void release() {
    try {
      encedec.stop();
      encedec.release();
      audioCapture.stop();
      audioCapture.release();
    } catch (Exception ignored) {
    }
  }
}

File: scrcpy_android/server/src/main/java/top/saymzx/scrcpy/server/helper/AudioEncode.kt
package top.saymzx.scrcpy.server.helper

import android.media.AudioRecord
import android.media.MediaCodec
import android.media.MediaFormat
import okio.BufferedSink
import top.saymzx.scrcpy.server.Server.isNormal
import top.saymzx.scrcpy.server.entity.Options

object AudioEncode {

  private lateinit var encedec: MediaCodec
  private lateinit var encodecFormat: MediaFormat
  private lateinit var audioCapture: AudioRecord
  lateinit var audioStream: BufferedSink

  fun stream(): Pair<Thread, Thread> {
    try {
      audioCapture = AudioCapture.start()
      setAudioEncodec()
      encedec.start()
    } catch (e: Exception) {
      print(e)
      isNormal = false
      try {
        audioCapture.stop()
        audioCapture.release()
        encedec.stop()
        encedec.release()
        audioStream.writeInt(0)
        return Pair(Thread {}, Thread {})
      } catch (_: Exception) {
      }
    }
    val threadIn = Thread {
      try {
        encodeIn()
      } catch (e: Exception) {
        print(e)
        isNormal = false
        try {
          audioCapture.stop()
          audioCapture.release()
          encedec.stop()
          encedec.release()
        } catch (_: Exception) {
        }
      }
    }
    threadIn.priority = Thread.MAX_PRIORITY
    val threadOut = Thread {
      try {
        encodeOut()
      } catch (e: Exception) {
        print(e)
        isNormal = false
        try {
          audioCapture.stop()
          audioCapture.release()
          encedec.stop()
          encedec.release()
        } catch (_: Exception) {
        }
      }
    }
    threadOut.priority = Thread.MAX_PRIORITY
    audioStream.writeInt(1)
    return Pair(threadIn, threadOut)
  }

  // 创建Codec
  private fun setAudioEncodec() {
    val codecMime =
      if (Options.audioCodec == "opus") MediaFormat.MIMETYPE_AUDIO_OPUS else MediaFormat.MIMETYPE_AUDIO_AAC
    encedec = MediaCodec.createEncoderByType(codecMime)
    encodecFormat = MediaFormat()
    encodecFormat.setString(MediaFormat.KEY_MIME, codecMime)
    encodecFormat.setInteger(MediaFormat.KEY_BIT_RATE, 64000)
    encodecFormat.setInteger(MediaFormat.KEY_CHANNEL_COUNT, AudioCapture.CHANNELS)
    encodecFormat.setInteger(MediaFormat.KEY_SAMPLE_RATE, AudioCapture.SAMPLE_RATE)
    encedec.configure(encodecFormat, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
  }

  // 编码入
  private fun encodeIn() {
    val size = AudioCapture.millisToBytes(5)
    var len: Int
    while (isNormal) {
      val inIndex = encedec.dequeueInputBuffer(-1)
      if (inIndex < 0) continue
      len = audioCapture.read(encedec.getInputBuffer(inIndex)!!, size)
      encedec.queueInputBuffer(inIndex, 0, len, 0, 0)
    }
  }

  // 编码出
  private fun encodeOut() {
    var outIndex: Int
    val bufferInfo = MediaCodec.BufferInfo()
    while (isNormal) {
      // 找到已完成的输出缓冲区
      outIndex = encedec.dequeueOutputBuffer(bufferInfo, -1)
      if (outIndex < 0) continue
      val buffer = encedec.getOutputBuffer(outIndex)
      audioStream.apply {
        writeInt(bufferInfo.size)
        writeLong(System.currentTimeMillis())
        write(buffer)
        flush()
      }
      encedec.releaseOutputBuffer(outIndex, false)
    }
  }
}